File: hydrograph.engine/hydrograph.engine.cascading/src/main/java/hydrograph/engine/cascading/integration/HydrographRuntime.java
/*******************************************************************************
 * Copyright 2017 Capital One Services, LLC and Bitwise, Inc.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License
 *******************************************************************************/
package hydrograph.engine.cascading.integration;

import cascading.cascade.Cascade;
import cascading.flow.Flow;
import cascading.property.AppProps;
import hydrograph.engine.component.mapping.ComponentAdapterFactory;
import hydrograph.engine.core.core.HydrographJob;
import hydrograph.engine.core.core.HydrographRuntimeService;
import hydrograph.engine.core.flowmanipulation.FlowManipulationContext;
import hydrograph.engine.core.flowmanipulation.FlowManipulationHandler;
import hydrograph.engine.core.helper.JAXBTraversal;
import hydrograph.engine.core.props.PropertiesLoader;
import hydrograph.engine.core.schemapropagation.SchemaFieldHandler;
import hydrograph.engine.core.utilities.CommandLineOptionsProcessor;
import hydrograph.engine.core.utilities.GeneralUtilities;
import hydrograph.engine.flow.utils.ExecutionTrackingListener;
import hydrograph.engine.hadoop.utils.HadoopConfigProvider;
import hydrograph.engine.jaxb.commontypes.TypeProperties.Property;
import hydrograph.engine.utilities.ExecutionTrackingUtilities;
import hydrograph.engine.utilities.HiveMetastoreTokenProvider;
import hydrograph.engine.utilities.UserClassLoader;
import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.util.GenericOptionsParser;
import org.apache.thrift.TException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.util.Properties;

@SuppressWarnings({"rawtypes"})
public class HydrographRuntime implements HydrographRuntimeService {

    final String EXECUTION_TRACKING = "hydrograph.execution.tracking";
    private static final String OPTION_DOT_PATH = "dotpath";
    private Properties hadoopProperties = new Properties();
    private ExecutionTrackingListener executionTrackingListener;
    private FlowBuilder flowBuilder;
    private RuntimeContext runtimeContext;
    private String[] args;
    private PropertiesLoader config;
    private FlowManipulationContext flowManipulationContext;
    private static Logger LOG = LoggerFactory.getLogger(HydrographRuntime.class);

    public void executeProcess(String[] args, HydrographJob hydrographJob) {
        this.args = args != null ? args.clone() : null;
        config = PropertiesLoader.getInstance();
        LOG.info("Invoking initialize on runtime service");
        initialize(config.getRuntimeServiceProperties(), this.args, hydrographJob, null,
                null);
        LOG.info("Preparation started");
        prepareToExecute();
        LOG.info("Preparation completed. Now starting execution");
        LOG.info("Execution Started");
        execute();
        LOG.info("Execution Complete");
        oncomplete();
    }

    public void initialize(Properties config, String[] args, HydrographJob hydrographJob, String jobId, String UDFPath) {

        AppProps.setApplicationName(hadoopProperties, hydrographJob.getJAXBObject().getName());

        hadoopProperties.putAll(config);

        Configuration conf = new HadoopConfigProvider(hadoopProperties).getJobConf();

        SchemaFieldHandler schemaFieldHandler = new SchemaFieldHandler(
                hydrographJob.getJAXBObject().getInputsOrOutputsOrStraightPulls());

        flowManipulationContext = new FlowManipulationContext(hydrographJob, args, schemaFieldHandler,
                jobId);

        FlowManipulationHandler flowManipulationHandler = new FlowManipulationHandler();

        hydrographJob = flowManipulationHandler.execute(flowManipulationContext);

        if (hydrographJob.getJAXBObject().getRuntimeProperties() != null
                && hydrographJob.getJAXBObject().getRuntimeProperties().getProperty() != null) {
            for (Property property : hydrographJob.getJAXBObject().getRuntimeProperties().getProperty()) {
                hadoopProperties.put(property.getName(), property.getValue());
            }
        }

        JAXBTraversal traversal = new JAXBTraversal(hydrographJob.getJAXBObject());

        if (traversal.isHiveComponentPresentInFlow()) {
            try {
                HiveMetastoreTokenProvider.obtainTokenForHiveMetastore(conf);
            } catch (TException e) {
                throw new HydrographRuntimeException(e);
            } catch (IOException e) {
                throw new HydrographRuntimeException(e);
            }
        }

        String[] otherArgs;
        try {
            otherArgs = new GenericOptionsParser(conf, args).getRemainingArgs();
        } catch (IOException e) {
            throw new HydrographRuntimeException(e);
        }

        String argsString = "";
        for (String arg : otherArgs) {
            argsString = argsString + " " + arg;
        }
        LOG.info("After processing arguments are:" + argsString);
        this.args = otherArgs;
        // setJar(otherArgs);

        hadoopProperties.putAll(conf.getValByRegex(".*"));

        ComponentAdapterFactory componentAdapterFactory = new ComponentAdapterFactory(hydrographJob.getJAXBObject());

        flowBuilder = new FlowBuilder();

        runtimeContext = new RuntimeContext(hydrographJob, traversal, hadoopProperties, componentAdapterFactory,
                flowManipulationContext.getSchemaFieldHandler(), UDFPath);

        LOG.info(
                "Graph '" + runtimeContext.getHydrographJob().getJAXBObject().getName() + "' initialized successfully");
    }

    @Override
    public void prepareToExecute() {
        flowBuilder.buildFlow(runtimeContext);

        if (GeneralUtilities.IsArgOptionPresent(args, OPTION_DOT_PATH)) {
            writeDotFiles();
        }

    }

    @Override
    public void execute() {
        if (GeneralUtilities.IsArgOptionPresent(args, CommandLineOptionsProcessor.OPTION_NO_EXECUTION)) {
            LOG.info(CommandLineOptionsProcessor.OPTION_NO_EXECUTION + " option is provided so skipping execution");
            return;
        }
        if (ExecutionTrackingUtilities.getExecutionTrackingClass(EXECUTION_TRACKING) != null) {
            executionTrackingListener = (ExecutionTrackingListener) UserClassLoader.loadAndInitClass(
                    ExecutionTrackingUtilities.getExecutionTrackingClass(EXECUTION_TRACKING), "execution tracking");
            executionTrackingListener.addListener(runtimeContext);
        }
        for (Cascade cascade : runtimeContext.getCascade()) {
            cascade.complete();
        }
    }

    @Override
    public void oncomplete() {
        flowBuilder.cleanup(flowManipulationContext.getTmpPath(), runtimeContext);
    }

    /**
     * Returns the statistics of components in a job.
     *
     * @see hydrograph.engine.execution.tracking.ComponentInfo
     */
    @Override
    public Object getExecutionStatus() {
        if (executionTrackingListener != null)
            return executionTrackingListener.getStatus();
        return null;
    }

    public Cascade[] getFlow() {
        return runtimeContext.getCascadingFlows();
    }

    private void writeDotFiles() {

        String[] paths = GeneralUtilities.getArgsOption(args, OPTION_DOT_PATH);

        if (paths == null) {
            throw new HydrographRuntimeException(
                    OPTION_DOT_PATH + " option is provided but is not followed by path");
        }
        String basePath = paths[0];
        LOG.info("Dot files will be written under " + basePath);

        String flowDotPath = basePath + "/" + runtimeContext.getHydrographJob().getJAXBObject().getName() + "/"
                + "flow";
        String flowStepDotPath = basePath + "/" + runtimeContext.getHydrographJob().getJAXBObject().getName() + "/"
                + "flowstep";

        int batchCounter = 0;
        for (Cascade cascadingFlow : runtimeContext.getCascadingFlows()) {
            for (Flow flows : cascadingFlow.getFlows()) {
                flows.writeDOT(flowDotPath + "_" + batchCounter);
                flows.writeStepsDOT(flowStepDotPath + "_" + batchCounter);
                batchCounter++;
            }
        }
    }


    public class HydrographRuntimeException extends RuntimeException {

        /**
         *
         */
        private static final long serialVersionUID = -7891832980227676974L;

        public HydrographRuntimeException(String msg) {
            super(msg);
        }

        public HydrographRuntimeException(Throwable e) {
            super(e);
        }
    }

    /**
     * Method to kill the job
     */
    @Override
    public void kill() {
        LOG.info("Kill signal received");
        if (runtimeContext.getCascade() != null) {
            for (Cascade cascade : runtimeContext.getCascade()) {
                LOG.info("Killing Cascading jobs: " + cascade.getID());
                cascade.stop();
            }
        } else
            LOG.info("No cascading jobs present to kill. Exiting code.");
        System.exit(0);
    }
}

File: hydrograph.engine/hydrograph.engine.spark/src/main/scala/hydrograph/engine/spark/flow/HydrographRuntime.scala
/*******************************************************************************
 * Copyright 2017 Capital One Services, LLC and Bitwise, Inc.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *******************************************************************************/
package hydrograph.engine.spark.flow

import java.io.IOException
import java.util.Properties

import hydrograph.engine.core.core.{HydrographJob, HydrographRuntimeService}
import hydrograph.engine.core.flowmanipulation.{FlowManipulationContext, FlowManipulationHandler}
import hydrograph.engine.core.helper.JAXBTraversal
import hydrograph.engine.core.props.OrderedProperties
import hydrograph.engine.core.schemapropagation.SchemaFieldHandler
import hydrograph.engine.core.utilities.OrderedPropertiesHelper
import hydrograph.engine.spark.components.adapter.factory.AdapterFactory
import hydrograph.engine.spark.components.base.SparkFlow
import hydrograph.engine.spark.executiontracking.plugin.{CommandComponentsDefaultPlugin, ExecutionTrackingListener, ExecutionTrackingPlugin, HydrographCommandListener}
import org.apache.hadoop.fs.{FileSystem, Path}
import org.apache.spark.SparkConf
import org.apache.spark.sql.SparkSession
import org.slf4j.{Logger, LoggerFactory}

import scala.collection.JavaConverters._
import scala.collection.mutable
import scala.util.matching.Regex

/**
  * The Class HydrographRuntime.
  *
  * @author Bitwise
  *
  */

class HydrographRuntime extends HydrographRuntimeService {

  private val EXECUTION_TRACKING: String = "hydrograph.execution.tracking"
  private val LOG: Logger = LoggerFactory.getLogger(classOf[HydrographRuntime])
  private var flowManipulationContext: FlowManipulationContext = null;
  //  private var flows: mutable.LinkedHashSet[SparkFlow] = null
  private var flowBuilder:FlowBuilder = null
  var executionTrackingListener : ExecutionTrackingListener = null
  var hydrographListener : HydrographCommandListener = new CommandComponentsDefaultPlugin


  override def kill(): Unit = {
    LOG.info("Kill signal received")
    if (RuntimeContext.instance.sparkSession != null) {
      LOG.info("Killing Spark jobs")
      RuntimeContext.instance.sparkSession.stop()
    }
    else {
      LOG.info("No Spark jobs present to kill. Exiting code.")
      System.exit(0)
    }
  }

  override def initialize(properties: Properties, args: Array[String], hydrographJob: HydrographJob,
                          jobId: String, udfPath: String): Unit = {

    val configProperties = getSparkProperties(hydrographJob,properties)

    val sparkSessionBuilder: SparkSession.Builder = SparkSession.builder()
      .appName(hydrographJob.getJAXBObject.getName)
      .config(configProperties)


    val schemaFieldHandler = new SchemaFieldHandler(
      hydrographJob.getJAXBObject().getInputsOrOutputsOrStraightPulls())

    val flowManipulationContext = new FlowManipulationContext(hydrographJob, args, schemaFieldHandler, jobId)

    val flowManipulationHandler = new FlowManipulationHandler

    val  updatedHydrographJob=flowManipulationHandler.execute(flowManipulationContext)

    val adapterFactory = AdapterFactory(updatedHydrographJob.getJAXBObject)

    val traversal = new JAXBTraversal(updatedHydrographJob.getJAXBObject());

    val sparkSession: SparkSession = enableHiveSupport(sparkSessionBuilder, traversal, properties).getOrCreate()

    getAndSetHadoopProperties(sparkSession,hydrographJob,properties)

    val runtimeContext = RuntimeContext(adapterFactory, traversal, updatedHydrographJob,
      flowManipulationContext.getSchemaFieldHandler, sparkSession)



    val EXECUTION_TRACKING = "hydrograph.execution.tracking";

 //        val oproperties = OrderedPropertiesHelper.getOrderedProperties("RegisterPlugin.properties")
 //        val executionTrackingPluginName = oproperties.getProperty(EXECUTION_TRACKING)
 //        val trackingInstance = Class.forName(executionTrackingPluginName).newInstance()
 //        executionTrackingListener = trackingInstance.asInstanceOf[ExecutionTrackingPlugin]
 //        executionTrackingListener.addListener(runtimeContext)


    if (getExecutionTrackingClass(EXECUTION_TRACKING) != null) {
      /*var */executionTrackingListener = classLoader(getExecutionTrackingClass(EXECUTION_TRACKING)).asInstanceOf[ExecutionTrackingListener]
      val trackingInstance = Class.forName(getExecutionTrackingClass(EXECUTION_TRACKING)).newInstance()
      executionTrackingListener = trackingInstance.asInstanceOf[ExecutionTrackingPlugin]
      executionTrackingListener.addListener(runtimeContext)
      hydrographListener = trackingInstance.asInstanceOf[ExecutionTrackingPlugin]
    }

  }

  def enableHiveSupport(sessionBuilder: SparkSession.Builder, traversal: JAXBTraversal, properties: Properties): SparkSession.Builder = {
    LOG.trace("In method checkAndEnableHiveSupport()")
    if (traversal.isHiveComponentPresentInFlow) {
      LOG.debug("Hive components are present in flow. Enabling Hive support in SparkSession with warehouse location "+properties.getProperty("hydrograph.hive.warehouse"))
      sessionBuilder
        .config("spark.sql.warehouse.dir", properties.getProperty("hydrograph.hive.warehouse"))
        .enableHiveSupport()
    } else {
      sessionBuilder
        .config("spark.sql.warehouse.dir", properties.getProperty("hydrograph.tmp.warehouse"))
    }
    sessionBuilder
  }

  override def prepareToExecute(): Unit = {
    LOG.info("Building spark flows")
    flowBuilder = FlowBuilder(RuntimeContext.instance, hydrographListener)
    LOG.info("Spark flows built successfully")
  }

  override def execute(): Unit = {
    /*if (GeneralUtilities.IsArgOptionPresent(args, CommandLineOptionsProcessor.OPTION_NO_EXECUTION)) {
      LOG.info(CommandLineOptionsProcessor.OPTION_NO_EXECUTION + " option is provided so skipping execution")
      return
    }*/
    /*for (sparkFlow <- flows) {
        try{
          hydrographListener.start(sparkFlow)
          //        HydrographFlowPlugin.getComps()
          sparkFlow.execute()
          hydrographListener.end(sparkFlow)
          /* for(accumulator <- sparkFlow.getAccumulatorOnFlow()){
             accumulator.reset()
           }*/
        }
        catch{case e: Exception => {
          hydrographListener.failComponentsOfFlow(sparkFlow)
          //                executionTrackingListener.getStatus().asScala.foreach(println)
          throw e
        }
      }
    }*/

    flowBuilder.buildAndExecuteFlows()
    //    RuntimeContext.instance.sparkSession.sparkContext.longAccumulator
    RuntimeContext.instance.sparkSession.stop()
    //        executionTrackingListener.getStatus().asScala.foreach(println)
  }

  override def getExecutionStatus: AnyRef = {
    if (executionTrackingListener != null)
      return executionTrackingListener.getStatus()
    return null
  }

  override def oncomplete(): Unit = {
    //Deleting TempPath For Debug
    if (flowManipulationContext != null && flowManipulationContext.getTmpPath != null) {
      flowManipulationContext.getTmpPath.asScala.foreach(tmpPath => {
        val fullPath: Path = new Path(tmpPath)
        // do not delete the root directory
        if (fullPath.depth != 0) {
          var fileSystem: FileSystem = null
          LOG.info("Deleting temp path:" + tmpPath)
          try {
            fileSystem = FileSystem.get(RuntimeContext.instance.sparkSession.sparkContext.hadoopConfiguration)
            //            fileSystem.delete(fullPath, true)
          }
          catch {
            case exception: NullPointerException => {
              throw new RuntimeException(exception)
            }
            case e: IOException => {
              throw new RuntimeException(e)
            }
          }
        }
      })
    }
  }

  def classLoader[T](className: String): T = {
    val clazz = Class.forName(className).getDeclaredConstructors
    clazz(0).setAccessible(true)
    clazz(0).newInstance().asInstanceOf[T]
  }

  def getSparkProperties(hydrographJob: HydrographJob, properties: Properties): SparkConf = {
    val configProperties = new SparkConf()
    val sparkProperties = properties.entrySet().asScala
    for (property <- sparkProperties) {
      if(property.getKey.toString.startsWith("spark.") || property.getKey.toString.startsWith("hydrograph.")){
        configProperties.set(property.getKey.toString.trim, property.getValue.toString.trim)
      }
    }
    val runttimeProperties = hydrographJob.getJAXBObject().getRuntimeProperties
    if (runttimeProperties != null) {
      for (runtimeProperty <- runttimeProperties.getProperty.asScala) {
        if(runtimeProperty.getName.startsWith("spark.") || runtimeProperty.getName.startsWith("hydrograph.")){
          configProperties.set(runtimeProperty.getName.trim, runtimeProperty.getValue.trim)
        }
      }
    }

    configProperties
  }

  def getAndSetHadoopProperties(sparkSession: SparkSession, hydrographJob: HydrographJob, properties: Properties): Unit = {

    val configProperties = sparkSession.sparkContext.hadoopConfiguration
    val sparkProperties = properties.entrySet().asScala
    for (property <- sparkProperties) {

      property.getKey.toString match {
        case a if(!a.startsWith("spark.")) => {
          val pattern: Regex = "(.*)\\((.*)\\)".r
          val matchIterator = pattern.findAllIn(property.getValue.toString)
          if (matchIterator.hasNext) {
            while (matchIterator.hasNext) {
              matchIterator.group(1) match {
                case x if (x.endsWith("OrElse")) => {
                  val arrayOfParams: Array[String] = matchIterator.group(2).split(",")
                  configProperties.set(property.getKey.toString.trim, sys.env.getOrElse(arrayOfParams(0).trim.replaceAll("\"", ""), arrayOfParams(1).trim.replaceAll("\"", "")).trim)
                }
                case y if (y.endsWith("get")) => {
                  configProperties.set(property.getKey.toString.trim, sys.env.get(matchIterator.group(2).replaceAll("\"", "")).get.trim)
                }
              }
              matchIterator.next()
            }
          }
          else{
            configProperties.set(property.getKey.toString.trim, property.getValue.toString.trim)
          }
        }
        case _ =>
      }

    }
    val runttimeProperties = hydrographJob.getJAXBObject().getRuntimeProperties
    if (runttimeProperties != null) {
      for (runtimeProperty <- runttimeProperties.getProperty.asScala) {
        runtimeProperty.getName match {
          case a if(!a.startsWith("spark.")) => {
            val pattern: Regex = "(.*)\\((.*)\\)".r
            val matchIterator = pattern.findAllIn(runtimeProperty.getValue)
            if (matchIterator.hasNext) {
              while (matchIterator.hasNext) {
                matchIterator.group(1) match {
                  case x if (x.endsWith("OrElse")) => {
                    val arrayOfParams: Array[String] = matchIterator.group(2).split(",")
                    configProperties.set(runtimeProperty.getName.trim, sys.env.getOrElse(arrayOfParams(0).replaceAll("\"", ""), arrayOfParams(1).replaceAll("\"", "")).trim)
                  }
                  case y if (y.endsWith("get")) => {
                    configProperties.set(runtimeProperty.getName.trim, sys.env.get(matchIterator.group(2).replaceAll("\"", "")).get.trim)

                  }
                }
                matchIterator.next()
              }
            }
            else{
              configProperties.set(runtimeProperty.getName.trim, runtimeProperty.getValue.trim)
            }
          }
          case _ =>
        }
      }
    }
  }


  def getExecutionTrackingClass(executionTrackingKey: String): String = {
    var properties: OrderedProperties = new OrderedProperties
    try {
      properties = OrderedPropertiesHelper.getOrderedProperties("RegisterPlugin.properties")
    }
    catch {

      case e: IOException => {
        throw new RuntimeException("Error reading the properties file: RegisterPlugin.properties" , e)
      }
    }
    properties.getProperty(executionTrackingKey)
  }
}